### PR TITLE
[Perf][Qwen3-Omni] code2wav: opportunistic micro-batching to remove global GPU lock

### PR DESCRIPTION
**Pull request by knitcapcat-amd**
## Motivation

Qwen3-Omni's `_Code2WavStreamingExecutor` serializes every vocoder
forward across concurrent requests through a global `asyncio.Lock`,
which is called out in #276 and is the "for higher throughput,
consider CUDA streams or batched inference" TODO left in the file's
own code comment. The vocoder is a ~200 MB CNN that naturally
supports batched inference, and per-forward wall time is dominated
by kernel launch / dispatch rather than compute.

Replacing the lock with an opportunistic micro-batch scheduler yields
**up to 8.93× throughput at concurrency=16** on MI300X with no TTFA
regression at concurrency=1.

## Modifications

This PR only changes files under `sglang_omni/models/qwen3_omni/` and
adds new files under `tests/` and `benchmarks/`; no other model's
code path is touched.

1. **Opportunistic micro-batch scheduler.** Per-request tasks submit
   `_DecodeRequest` objects to a shared queue; a dedicated batch loop
   drains everything currently pending, pads to common length, runs
   one batched forward, and dispatches per-request results through
   Futures. Single-request batches take an unchanged fast path.
2. **Pad-value correctness fix.** `_forward_batch` used to pad with
   `codec_eos_token_id=2150`, which is outside code2wav's codebook
   range `[0, 2048)` — any variable-length batch would have triggered
   a CUDA illegal access inside `code_embedding(codes + offset)`. The
   bug was hidden by the old fixed-length scheduler. Introduced
   `pad_token_id` (default `0`) decoupled from the EOS sentinel.
3. **MIOpen / cuDNN warmup at construction.** First forward at each
   new batch size JIT-compiles a conv solver (~4.5 s per shape on
   ROCm). Primes `{1, 2, 4, 8, max_batch_size}` up front. Opt-out
   via `warmup=False`.
4. **Resilient batch loop.** Inner step sets exceptions on in-flight
   Futures before re-raising; outer supervisor swallows non-cancel
   exceptions so a one-off CUDA OOM can't deadlock callers.

## Related Issues

Refs #276 (Phase 1 of a multi-stage talker pipeline optimization;
Phase 2 on `code_predictor` and Phase 3 on Talker-AR HOL blocking to
follow as separate PRs).

## Accuracy Test

Vocoder model is unchanged; 8 unit tests in
`tests/test_code2wav_executor.py` cover the executor changes
(`pytest tests/test_code2wav_executor.py --asyncio-mode=auto` — all 8
pass). Two keep me honest:

- `test_forward_batch_matches_independent_single_forwards`: builds a
  variable-length batch and asserts bit-identical waveforms vs running
  each request through the single-forward path.
- `test_forward_batch_rejects_out_of_range_pad_token_id`: regression
  test that fails if anyone reverts pad to an OOB id.

The mock vocoder goes through a real `nn.Embedding`, so OOB ids fail
tests the same way the real `Qwen3OmniMoeCode2Wav` does.

## Benchmark & Profiling

1× MI300X, Qwen3-Omni-30B-A3B-Instruct (code2wav only), bfloat16, 32
requests per level, `chunks_per_request=10`, `left_context_size=25`
(production default). Driver:
`benchmarks/perf/bench_code2wav.py` on both branches through the
executor's public `add_request` + `StreamQueue` API. Random RVQ codes
(`low=0, high=2048`).

| Concurrency | main wall (ms) | this PR wall (ms) | main req/s | this PR req/s |   Speedup |
| :---------: | -------------: | ----------------: | ---------: | ------------: | --------: |
|      1      |            625 |               618 |      51.21 |         51.76 |    parity |
|      4      |            614 |               171 |      52.13 |        187.01 | **3.59×** |
|      8      |            617 |               102 |      51.87 |        312.41 | **6.02×** |
|     16      |            620 |                69 |      51.65 |        461.22 | **8.93×** |

`main` wall is flat in concurrency (lock-bound); `this PR` scales
monotonically as batching amortizes per-forward overhead. `c=1` parity
validates the fast path has no regression. The `c=16` number exceeds
the early plan expectation of 4–6× because the in-executor MIOpen
warmup eliminates conv-solver JIT cold start that would otherwise leak
into the measurement window.

Caveats: bench uses best-case all-at-once arrival
(`asyncio.gather` + `Semaphore(concurrency)`) so measured speedup is
an upper bound for this workload shape; steady-state padding overhead
is ≤ 10% (most decodes are already at `max_len = context + chunks =
35`).

### Note on E2E benchmarking

This PR is validated with a micro-benchmark against the code2wav
executor rather than an end-to-end TTS benchmark. Issue #276 and
#265's abandonment of the unified accuracy+WER+speed CI both identify
the `SGLangBatchPlanner` HOL blocking in talker AR decode as the
dominant bottleneck at `concurrency=8`. That is out of scope here
(future Phase 3) and cannot move meaningfully from a code2wav-only
change. The micro-benchmark follows the same methodology as
[vllm-omni PR #1246](https://github.com/vllm-project/vllm-omni/pull/1246),
which also justified its single-stage vocoder change with per-stage
data, so the two are directly comparable.

## Checklist

- [x] Format your code according with pre-commit.
- [x] Add unit tests.
- [x] Update documentation / docstrings / example tutorials as needed.
- [x] Provide throughput / latency benchmark results and accuracy
  evaluation results as needed.
- [ ] For reviewers: If you haven't made any contributions to this PR
  and are only assisting with merging the main branch, please remove
  yourself as a co-author when merging the PR.

### CODE REVIEW

File: benchmarks/perf/bench_code2wav.py [hunk 1]
```diff
@@ -0,0 +1,370 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (JingwenGu0829) [inline benchmarks/perf/bench_code2wav.py:1]: Hi, I don't think we need to include additional benchmark in this pr since the benchmark is usually testing the performance of the whole pr by shared server-level methods.

File: benchmarks/perf/bench_code_predictor.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (JingwenGu0829) [inline benchmarks/perf/bench_code_predictor.py:1]: Same us above, I don't think we need to introduce benchmarks in this pr.

File: sglang_omni/vendor/sglang/layers.py [hunk 3]
```diff
@@ -14,7 +14,14 @@
 from typing import Optional, Tuple, Union
 
 import torch
-from sgl_kernel import top_k_top_p_sampling_from_probs
+
+try:
```
Comment (JingwenGu0829) [inline sglang_omni/vendor/sglang/layers.py:18]: I don't think we should be using this try except import in development phase yet? I think this error could only caused by running with incorrect docker.

**Comment by JingwenGu0829** [comment]
This PR moves to #319
